addons/sale_project/models/project.py

```diff
@@ -233,6 +246,149 @@ def _get_sale_order_items_query(self, domain_per_model=None):
         query._where_params = project_params + task_params + milestone_params
         return query
 
+    def get_panel_data(self):
+        panel_data = super().get_panel_data()
+        return {
+            **panel_data,
+            'sale_items': self._get_sale_items(),
+        }
+
+    def get_sale_items_data(self, domain=None, offset=0, limit=None, with_action=True):
+        if not self.user_has_groups('project.group_project_user'):
+            return {}
+        sols = self.env['sale.order.line'].sudo().search(
+            domain or self._get_sale_items_domain(),
+            offset=offset,
+            limit=limit,
+        )
+        # filter to only get the action for the SOLs that the user can read
+        action_per_sol = sols._filter_access_rules_python('read')._get_action_per_item() if with_action else {}
+
+        def get_action(sol_id):
+            """ Return the action vals to call it in frontend if the user can access to the SO related """
+            action = action_per_sol.get(sol_id)
+            return {'action': action, 'additional_context': json.dumps({'active_id': sol_id})} if action else {}
+
+        return [{
+            **sol_read,
+            **get_action(sol_read['id']),
+        } for sol_read in sols.read(['display_name', 'product_uom_qty', 'qty_delivered', 'qty_invoiced', 'product_uom'])]
+
+    def _get_sale_items_domain(self, additional_domain=None):
+        sale_orders = self._get_sale_order_items().sudo().order_id
+        domain = [('order_id', 'in', sale_orders.ids), ('is_downpayment', '=', False), ('state', 'in', ['sale', 'done'])]
+        if additional_domain:
+            domain = expression.AND([domain, additional_domain])
+        return domain
+
+    def _get_sale_items(self, with_action=True):
+        domain = self.sudo()._get_sale_items_domain()
```

Comment:
In that case it is to avoid having the `ir.rule` for project. I assume in that method the current user has access to the project.

What do you think? 🙂

Reply:
The idea was that if the sudo is place before I did not understand why using it inside _get_sale_items_domain.
But you are right, in case _get_sale_items_domain is used elsewhere it is then needed.